aiida/backends/tests/work/workChain.py

```diff
@@ -385,6 +386,56 @@ def _run_with_checkpoints(self, wf_class, inputs=None):
 
         return finished_steps
 
+class TestFastForwardingWorkChain(TestWorkchain):
+    def setUp(self):
+        super(TestFastForwardingWorkChain, self).setUp()
+        class ReturnInputsFastForward(WorkChain):
+            @classmethod
+            def define(cls, spec):
+                super(ReturnInputsFastForward, cls).define(spec)
+                spec.input('a', valid_type=Int)
+                spec.input('b', valid_type=Int, default=Int(2), required=False)
+                spec.outline(cls.return_inputs)
+                spec.deterministic()
+
+            def return_inputs(self):
+                self.out('a', self.inputs.a)
+                self.out('b', self.inputs.b)
+        self.wf_class = ReturnInputsFastForward
+        self.reference_result, self.reference_pid = run(
+            self.wf_class, a=Int(1), b=Int(2), _return_pid=True
+        )
+        self.reference_wc = load_node(self.reference_pid)
+
+        class ReturnInputsFastForward(WorkChain):
+            @classmethod
+            def define(cls, spec):
+                super(ReturnInputsFastForward, cls).define(spec)
+                spec.input('a', valid_type=Int)
+                spec.input('b', valid_type=Int, default=Int(2), required=False)
+                spec.outline(cls.return_inputs)
+                spec.deterministic()
+
+            def return_inputs(self):
+                raise ValueError
+
+        self.wf_class_broken = ReturnInputsFastForward
+
+    def tearDown(self):
+        super(TestFastForwardingWorkChain, self).tearDown()
+        super(TestFastForwardingWorkChain, self).tearDownClass()
+        super(TestFastForwardingWorkChain, self).setUpClass()
+
+    def test_hash(self):
```

Comment:
I was a bit confused by this test as I remembered the discussion that `WorkCalculations` can not be cached, but then here there is actually a test that checks that the caching of `WorkChains` works by passing `_use_cache=True` as an input. Granted, it only checks whether the resulting hashes are equal, but I wondered whether the second workchain was actually taken from the cache or not.

After a lot of digging, I noticed that caching is disabled for `WorkCalculations` because the `_cacheable` property is set to `False` for the `AbstractCalculation` and then reenabled only for the `InlineCalculation` and `JobCalculation`. Additionally I found that you really enforce this behavior, by putting an extra safeguard in the `_store_from_cache` method of the `Node` class in `aiida.orm.implementation.general.node`, by checking whether the node has any `RETURN` links, in which case you don't use the cache, just in case people subclass and override the `_cacheable` property.

Ultimately, I think the current implementation results in the correct behavior, however, I think it would be good to describe this implementation detail in the developer's guide somewhere. It is too specialized for users, but I think it is very important knowledge for developers and it was a bit tricky to track down.

Reply:
I probably forgot to remove that test after deciding that caching doesn't work for ``WorkCalculations`` -- I deleted it now.

As for ``_cacheable`` and the ``RETURN`` links, I added a small section in the developer's guide.